Build still fails with gcc on a missing header: `eigen.h: error: 'assert' was not declared in this scope`

Compiling with gcc errs out due to a missing header, which I pointed out long ago and suggested a fix in https://github.com/artivis/manif/pull/277/commits/bd2d162fcb7524bf9b67ff9116bfe12918f2002f which was not accepted. From the current de52f8b8bf354e5e03a61025986f4fdb1abad774 the build still fails:
```
:info:build [  8%] Building CXX object test/so2/CMakeFiles/gtest_so2.dir/gtest_so2.cpp.o
:info:build cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/build/test/so2 && /opt/local/bin/g++-mp-14  -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/external/tl -isystem /opt/local/include/eigen3 -isystem /opt/local/include -pipe -Os -DNDEBUG -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -mmacosx-version-min=10.6 -Werror=all -Werror=extra -MD -MT test/so2/CMakeFiles/gtest_so2_tangent_map.dir/gtest_so2_tangent_map.cpp.o -MF CMakeFiles/gtest_so2_tangent_map.dir/gtest_so2_tangent_map.cpp.o.d -o CMakeFiles/gtest_so2_tangent_map.dir/gtest_so2_tangent_map.cpp.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/so2/gtest_so2_tangent_map.cpp
:info:build cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/build/test/so2 && /opt/local/bin/g++-mp-14  -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/external/tl -isystem /opt/local/include/eigen3 -isystem /opt/local/include -pipe -Os -DNDEBUG -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -mmacosx-version-min=10.6 -Werror=all -Werror=extra -MD -MT test/so2/CMakeFiles/gtest_so2_tangent.dir/gtest_so2_tangent.cpp.o -MF CMakeFiles/gtest_so2_tangent.dir/gtest_so2_tangent.cpp.o.d -o CMakeFiles/gtest_so2_tangent.dir/gtest_so2_tangent.cpp.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/so2/gtest_so2_tangent.cpp
:info:build cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/build/test && /opt/local/bin/g++-mp-14  -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/external/tl -isystem /opt/local/include/eigen3 -isystem /opt/local/include -pipe -Os -DNDEBUG -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -mmacosx-version-min=10.6 -Werror=all -Werror=extra -MD -MT test/CMakeFiles/gtest_misc.dir/gtest_misc.cpp.o -MF CMakeFiles/gtest_misc.dir/gtest_misc.cpp.o.d -o CMakeFiles/gtest_misc.dir/gtest_misc.cpp.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/gtest_misc.cpp
:info:build cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/build/test/so2 && /opt/local/bin/g++-mp-14  -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/external/tl -isystem /opt/local/include/eigen3 -isystem /opt/local/include -pipe -Os -DNDEBUG -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -mmacosx-version-min=10.6 -Werror=all -Werror=extra -MD -MT test/so2/CMakeFiles/gtest_so2.dir/gtest_so2.cpp.o -MF CMakeFiles/gtest_so2.dir/gtest_so2.cpp.o.d -o CMakeFiles/gtest_so2.dir/gtest_so2.cpp.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/so2/gtest_so2.cpp
:info:build In file included from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/lie_group_base.h:6,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/SO2.h:6,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/so2/gtest_so2.cpp:1:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/so2/SO2_base.h: In instantiation of 'Eigen::Matrix<typename manif::LieGroupBase<Derived>::Scalar, 2, 1> manif::SO2Base<_Derived>::act(const Eigen::MatrixBase<OtherDerived>&, tl::optional<Eigen::Ref<Eigen::Matrix<typename manif::LieGroupBase<Derived>::Scalar, 2, 1> > >, tl::optional<Eigen::Ref<Eigen::Matrix<typename manif::LieGroupBase<Derived>::Scalar, 2, 2> > >) const [with _EigenDerived = Eigen::Matrix<double, 2, 1>; _Derived = manif::SO2<double>; typename manif::LieGroupBase<Derived>::Scalar = double; typename std::conditional<Eigen::Matrix<typename manif::LieGroupBase<Derived>::Scalar, 2, 1>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::InnerStride<1>; typename std::conditional<Eigen::Matrix<typename manif::LieGroupBase<Derived>::Scalar, 2, 2>::IsVectorAtCompileTime, Eigen::InnerStride<1>, Eigen::OuterStride<> >::type = Eigen::OuterStride<>]':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/so2/gtest_so2.cpp:509:35:   required from here
:info:build   509 |   auto transformed_point = so2.act(Eigen::Vector2d(1,1));
:info:build       |                            ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:94:9: error: 'assert' was not declared in this scope
:info:build    94 |   assert(x.cols() == 1 && "Expected a vector !"); \
:info:build       |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:97:3: note: in expansion of macro 'assert_is_vector'
:info:build    97 |   assert_is_vector(x); \
:info:build       |   ^~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/so2/SO2_base.h:243:3: note: in expansion of macro 'assert_vector_dim'
:info:build   243 |   assert_vector_dim(v, 2);
:info:build       |   ^~~~~~~~~~~~~~~~~
:info:build In file included from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/SO2.h:10:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/so2/SO2_base.h:1:1: note: 'assert' is defined in header '<cassert>'; this is probably fixable by adding '#include <cassert>'
:info:build   +++ |+#include <cassert>
:info:build     1 | #ifndef _MANIF_MANIF_SO2_BASE_H_
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:64:9: error: 'assert' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation
:info:build    64 |   assert(x.rows() == dim && "x.rows != "#dim" .");
:info:build       |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:98:3: note: in expansion of macro 'assert_rows_dim'
:info:build    98 |   assert_rows_dim(x, dim);
:info:build       |   ^~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/so2/SO2_base.h:243:3: note: in expansion of macro 'assert_vector_dim'
:info:build   243 |   assert_vector_dim(v, 2);
:info:build       |   ^~~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:94:9: note: 'assert' declared here, later in the translation unit
:info:build    94 |   assert(x.cols() == 1 && "Expected a vector !"); \
:info:build       |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:97:3: note: in expansion of macro 'assert_is_vector'
:info:build    97 |   assert_is_vector(x); \
:info:build       |   ^~~~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/so2/SO2_base.h:243:3: note: in expansion of macro 'assert_vector_dim'
:info:build   243 |   assert_vector_dim(v, 2);
:info:build       |   ^~~~~~~~~~~~~~~~~
```

There is a new error with gcc-14 also:
```
:info:build [ 70%] Building CXX object test/bundle/CMakeFiles/gtest_bundle.dir/gtest_bundle.cpp.o
:info:build cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/build/test/bundle && /opt/local/bin/g++-mp-14  -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/external/tl -isystem /opt/local/include/eigen3 -isystem /opt/local/include -pipe -Os -DNDEBUG -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=c++14 -mmacosx-version-min=10.6 -Werror=all -Werror=extra -MD -MT test/bundle/CMakeFiles/gtest_bundle.dir/gtest_bundle.cpp.o -MF CMakeFiles/gtest_bundle.dir/gtest_bundle.cpp.o.d -o CMakeFiles/gtest_bundle.dir/gtest_bundle.cpp.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/gtest_bundle.cpp
:info:build In file included from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/sgal3/gtest_sgal3.cpp:4:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/sgal3/../common_tester.h: In instantiation of 'void manif::CommonTester<_LieGroup>::evalAvgBiInvariant() [with _LieGroup = manif::SGal3<double>]':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/sgal3/gtest_sgal3.cpp:356:1:   required from here
:info:build    85 |   { evalAvgBiInvariant(); }                                               \
:info:build       |     ~~~~~~~~~~~~~~~~~~^~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/sgal3/../common_tester.h:510:26: warning: 'void std::random_shuffle(_RAIter, _RAIter) [with _RAIter = __gnu_cxx::__normal_iterator<manif::SGal3<double>*, vector<manif::SGal3<double>, Eigen::aligned_allocator<manif::SGal3<double> > > >]' is deprecated: use 'std::shuffle' instead [-Wdeprecated-declarations]
:info:build   510 |       std::random_shuffle( mans.begin(), mans.end() );
:info:build       |       ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
:info:build In file included from /opt/local/include/gcc14/c++/algorithm:61,
:info:build                  from /opt/local/include/eigen3/Eigen/Core:92,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:6,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/lie_group_base.h:6,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/SGal3.h:5,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/sgal3/gtest_sgal3.cpp:3:
:info:build /opt/local/include/gcc14/c++/bits/stl_algo.h:4491:5: note: declared here
:info:build  4491 |     random_shuffle(_RandomAccessIterator __first, _RandomAccessIterator __last)
:info:build       |     ^~~~~~~~~~~~~~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/gtest_bundle.cpp: In member function 'virtual void Bundle_Interface_Test::TestBody()':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/gtest_bundle.cpp:103:27: error: unused variable 'vec' [-Werror=unused-variable]
:info:build   103 |   typename GroupA::Vector vec;
:info:build       |                           ^~~
:info:build In file included from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/gtest_bundle.cpp:5:
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/../common_tester.h: In instantiation of 'void manif::CommonTester<_LieGroup>::evalAvgBiInvariant() [with _LieGroup = manif::Bundle<double, manif::R2, manif::SO3, manif::R1>]':
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/gtest_bundle.cpp:202:1:   required from here
:info:build    85 |   { evalAvgBiInvariant(); }                                               \
:info:build       |     ~~~~~~~~~~~~~~~~~~^~
:info:build /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/../common_tester.h:510:26: warning: 'void std::random_shuffle(_RAIter, _RAIter) [with _RAIter = __gnu_cxx::__normal_iterator<manif::Bundle<double, manif::R2, manif::SO3, manif::R1>*, vector<manif::Bundle<double, manif::R2, manif::SO3, manif::R1>, Eigen::aligned_allocator<manif::Bundle<double, manif::R2, manif::SO3, manif::R1> > > >]' is deprecated: use 'std::shuffle' instead [-Wdeprecated-declarations]
:info:build   510 |       std::random_shuffle( mans.begin(), mans.end() );
:info:build       |       ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
:info:build In file included from /opt/local/include/gcc14/c++/algorithm:61,
:info:build                  from /opt/local/include/eigen3/Eigen/Core:92,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/eigen.h:6,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/impl/lie_group_base.h:6,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/functions.h:4,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/include/manif/manif.h:5,
:info:build                  from /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_devel_manif/manif/work/manif-de52f8b8bf354e5e03a61025986f4fdb1abad774/test/bundle/gtest_bundle.cpp:3:
:info:build /opt/local/include/gcc14/c++/bits/stl_algo.h:4491:5: note: declared here
:info:build  4491 |     random_shuffle(_RandomAccessIterator __first, _RandomAccessIterator __last)
:info:build       |     ^~~~~~~~~~~~~~
```

--- a/include/manif/impl/eigen.h
+++ b/include/manif/impl/eigen.h
@@ -1,6 +1,8 @@
 #ifndef _MANIF_MANIF_EIGEN_H_
 #define _MANIF_MANIF_EIGEN_H_
 
+#include <cassert>
+
 #include <Eigen/Core>
 #include <Eigen/LU> // for mat.inverse()
 #include <Eigen/Geometry>

--- a/test/bundle/gtest_bundle.cpp
+++ b/test/bundle/gtest_bundle.cpp
@@ -100,8 +100,6 @@ TEST(Bundle, Interface)
   EXPECT_EIGEN_NEAR(G_Ginv.element<1>().inverse().coeffs(), SO3d::Identity().coeffs());
   EXPECT_EIGEN_NEAR(G_Ginv.element<2>().inverse().coeffs(), R1d::Identity().coeffs());
 
-  typename GroupA::Vector vec;
-
   auto adj = G.adj();
   Eigen::Matrix2d adj0 = adj.block<2, 2>(0, 0);
   EXPECT_EIGEN_NEAR(adj0, G.element<0>().adj());

